src/Parsing/Impl/Ast/ScopeStatement.cs

```diff
@@ -13,276 +13,82 @@
-                    }
-
-                    if (variable.Kind == VariableKind.Local) {
-                        Debug.Assert(variable.Scope == this);
-                    }
-                }
-            }
-
-            // no longer needed
-            _references = null;
-        }
-
-        private static bool HasClosureVariable(List<ClosureInfo> closureVariables, PythonVariable variable) {
-            if (closureVariables == null) {
-                return false;
-            }
-
-            for (var i = 0; i < closureVariables.Count; i++) {
-                if (closureVariables[i].Variable == variable) {
-                    return true;
-                }
-            }
-
-            return false;
-        }
-
-        private void EnsureVariables() {
-            if (Variables == null) {
-                Variables = new Dictionary<string, PythonVariable>(StringComparer.Ordinal);
-            }
-        }
-
-        internal void AddVariable(PythonVariable variable) {
-            EnsureVariables();
-            Variables[variable.Name] = variable;
-        }
-
-        internal PythonReference Reference(string/*!*/ name) {
-            if (_references == null) {
-                _references = new Dictionary<string, List<PythonReference>>(StringComparer.Ordinal);
-            }
-
-            if (!_references.TryGetValue(name, out var references)) {
-                _references[name] = references = new List<PythonReference>();
-            }
-            var reference = new PythonReference(name);
-            references.Add(reference);
-            return reference;
-        }
-
-        internal bool IsReferenced(string name) => _references != null && _references.ContainsKey(name);
-
-        internal PythonVariable/*!*/ CreateVariable(string name, VariableKind kind) {
-            EnsureVariables();
-            PythonVariable variable;
-            Variables[name] = variable = new PythonVariable(name, kind, this);
-            return variable;
-        }
-
-        internal PythonVariable/*!*/ EnsureVariable(string/*!*/ name) {
-            if (!TryGetVariable(name, out var variable)) {
-                return CreateVariable(name, VariableKind.Local);
-            }
-            return variable;
-        }
-
-        internal PythonVariable DefineParameter(string name) => CreateVariable(name, VariableKind.Parameter);
-
-        struct ClosureInfo {
-            public PythonVariable Variable;
-            public bool AccessedInScope;
-
-            public ClosureInfo(PythonVariable variable, bool accessedInScope) {
-                Variable = variable;
-                AccessedInScope = accessedInScope;
-            }
-        }
+        public IReadOnlyList<PythonVariable> FreeVariables => ScopeDelegate.FreeVariables;
+        #endregion
+
+        #region IBindableNode
+        void IBindableNode.Bind(PythonNameBinder binder) => ScopeDelegate.Bind(binder);
```

Comment:
Normally we don't use explicit interface implementations. Is there anything that conflicts with the interface members?

Reply:
I read that when we have a public class that implements an internal interface, the methods from the internal interface are still available. I tried this by checking out some of the methods from the `LanguageServer` project and I could still access those methods. When I did explicit implementation then those methods were hidden. If you want I could switch them back to implicit implementation.

Reply:
That's a good observation.

Reply:
if class is public and method is public the method is, of course, accessible. Having it on an internal interface does not change this fact.

Reply:
I think that's what he's saying, but by explicitly implementing the interface he's not actually exporting those names unless the class is used as the interface. (i.e. https://stackoverflow.com/a/6117670 but in our case the class is public, not internal)

The only other way I can think of to keep "IBindableNode" completely internal would be to have some internal abstract class that implements IBindableNode which each of these classes inherit from and complete...

Reply:
Ah, OK